### PR TITLE
Improve performance of language detection

### LINKED ISSUE #129576: Automatic language detection telemetry
https://github.com/microsoft/vscode/issues/129436

For telemetry we should capture the following:

1. What `language` we choose with what `confidence` and how long the `computation` took.
2. If the user explicitly set a different language after we automatically picked (this would be a signal that we were wrong)

This telemetry will give us good insights and confidence for the future when we decide how aggressively should we expose this feature.

### PR DESCRIPTION
**Pull request by TylerLeonhardt**
partially addresses #129607 

Broken up by commit:

1. moves everything into a worker
2. moves event handling to UntitledEditorModel (more code organization than anything else...)
3. trying to reuse as many interfaces and classes in the worker
4. adds telemetry fixes #129576

### CODE REVIEW

File: src/buildfile.js [hunk 1]
```diff
@@ -17,6 +17,7 @@ exports.base = [{
 
 exports.workerExtensionHost = [entrypoint('vs/workbench/services/extensions/worker/extensionHostWorker')];
 exports.workerNotebook = [entrypoint('vs/workbench/contrib/notebook/common/services/notebookSimpleWorker')];
+exports.workerLanguageDetection = [entrypoint('vs/workbench/services/languageDetection/common/languageDetectionSimpleWorker')];
```
Comment (TylerLeonhardt) [inline src/buildfile.js:20]: Is there a way I could include the `@vscode/vscode-languagedetection` npm package here?

File: src/vs/workbench/services/languageDetection/browser/languageDetectionWorkerServiceImpl.ts [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+		@IConfigurationService private readonly _configurationService: IConfigurationService,
+		@IModelService modelService: IModelService,
+	) {
+		super();
+
+		this._languageDetectionWorkerClient = new LanguageDetectionWorkerClient(
+			modelService,
+			'languageDetectionWorkerService',
+			// TODO: See if it's possible to bundle vscode-languagedetection
+			FileAccess.asBrowserUri('../../../../../../node_modules/@vscode/vscode-languagedetection/dist/lib/index.js', require).toString(true),
```
Comment (TylerLeonhardt) [inline src/vs/workbench/services/languageDetection/browser/languageDetectionWorkerServiceImpl.ts:40]: The worker had trouble importing this module so I had to pass a Uri to it. There's gotta be a better way....

File: src/vs/workbench/services/languageDetection/browser/languageDetectionWorkerServiceImpl.ts [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+		return languages.filter(<T>(l?: T): l is T => Boolean(l));
+	}
+}
+
+export interface IWorkerClient<W> {
+	getProxyObject(): Promise<W>;
+	dispose(): void;
+}
+
+class LanguageDetectionModelManager extends Disposable {
```
Comment (TylerLeonhardt) [inline src/vs/workbench/services/languageDetection/browser/languageDetectionWorkerServiceImpl.ts:84]: This was lifted from here:
https://github.com/microsoft/vscode/blob/025c9ce85fd369e668154c817d58098428cfe4b2/src/vs/editor/common/services/editorWorkerServiceImpl.ts#L276-L276

can I reuse it somehow?

File: src/vs/workbench/services/languageDetection/browser/languageDetectionWorkerServiceImpl.ts [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+				this._workerFactory,
+				'vs/workbench/services/languageDetection/browser/languageDetectionSimpleWorker',
+				new LanguageDetectionWorkerHost(
+					this.indexJsUri,
+					this.modelJsonUri,
+					this.weightsUri)
+			));
+		}
+		return this._worker;
+	}
```
Comment (TylerLeonhardt) [inline src/vs/workbench/services/languageDetection/browser/languageDetectionWorkerServiceImpl.ts:238]: mostly lifted from https://github.com/microsoft/vscode/blob/025c9ce85fd369e668154c817d58098428cfe4b2/src/vs/editor/common/services/editorWorkerServiceImpl.ts#L428-L457

can I reuse somehow?

File: src/vs/workbench/services/languageDetection/browser/languageDetectionSimpleWorker.ts [hunk 5]
```diff
@@ -0,0 +1,455 @@
+/*---------------------------------------------------------------------------------------------
```
Comment (TylerLeonhardt) [inline src/vs/workbench/services/languageDetection/browser/languageDetectionSimpleWorker.ts:1]: Loads of this file is copied... I should try to reuse instead.

**Comment by TylerLeonhardt** [comment]
So I noticed that this 404s:
image

because it's the incorrect path... it makes be believe that the `import()` inside of the worker isn't honoring the paths defined here:
https://github.com/microsoft/vscode/blob/d7de341c8d7e3e1ebab712ea139f32c70ae54a06/src/vs/code/browser/workbench/workbench-dev.html#L45

(or in the prod version)

So I wonder if I could at least query that path somehow in the UI thread and then I can pass that down to the worker...

**Comment by TylerLeonhardt** [comment]
Using the same logic as the model.js and bin file, I was able to get the vscode-languagedetection js file loaded in the worker both in Desktop and web. Marking as ready for review.

**Comment by TylerLeonhardt** [comment]
@alexdima, I will need to talk to you when you get back about whether this is the right idea. I think it would be better to instead bundle my npm module into the worker.

File: src/vs/workbench/services/languageDetection/browser/languageDetectionWorkerServiceImpl.ts [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+		@ITelemetryService telemetryService: ITelemetryService,
+	) {
+		super();
+
+		this._languageDetectionWorkerClient = new LanguageDetectionWorkerClient(
+			modelService,
+			telemetryService,
+			'languageDetectionWorkerService',
+			// TODO: See if it's possible to bundle vscode-languagedetection
+			this._environmentService.isBuilt && !isWeb
```
Comment (bpasero) [inline src/vs/workbench/services/languageDetection/browser/languageDetectionWorkerServiceImpl.ts:43]: This would probably read easier if the common prefix up until `vscode-languagedetection` was extracted to a variable.

File: src/vs/workbench/services/languageDetection/browser/languageDetectionWorkerServiceImpl.ts [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+		return languages.filter(<T>(l?: T): l is T => Boolean(l));
+	}
+}
+
+export interface IWorkerClient<W> {
+	getProxyObject(): Promise<W>;
+	dispose(): void;
+}
+
+class LanguageDetectionModelManager extends Disposable {
```
Comment (bpasero) [inline src/vs/workbench/services/languageDetection/browser/languageDetectionWorkerServiceImpl.ts:89]: I share the sentiment that this is code duplication for no good reason.

File: src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts [hunk 8]
```diff
@@ -368,6 +389,19 @@ export class UntitledTextEditorModel extends BaseTextEditorModel implements IUnt
 
 		// Emit as general content change event
 		this._onDidChangeContent.fire();
+
+		this.detectLanguageIfEnabled();
+	}
+
+	@debounce(600)
+	private async detectLanguageIfEnabled() {
```
Comment (bpasero) [inline src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts:397]: Maybe just `autoDetectLanguage`

File: src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts [hunk 9]
```diff
@@ -368,6 +389,19 @@ export class UntitledTextEditorModel extends BaseTextEditorModel implements IUnt
...
+		if (this.hasModeSetExplicitly || !this.languageDetectionService.isEnabledForMode(this.getMode() ?? '')) {
+			return;
+		}
+
+		const lang = await this.languageDetectionService.detectLanguage(this.resource);
+		if (!lang) { return; }
```
Comment (bpasero) [inline src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts:403]: My personal preference is to not put this into a single line.

File: src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts [hunk 10]
```diff
@@ -368,6 +389,19 @@ export class UntitledTextEditorModel extends BaseTextEditorModel implements IUnt
 
 		// Emit as general content change event
 		this._onDidChangeContent.fire();
+
+		this.detectLanguageIfEnabled();
+	}
+
+	@debounce(600)
+	private async detectLanguageIfEnabled() {
+		if (this.hasModeSetExplicitly || !this.languageDetectionService.isEnabledForMode(this.getMode() ?? '')) {
```
Comment (bpasero) [inline src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts:398]: I do not think a empty string exists as mode?

File: src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts [hunk 11]
```diff
@@ -368,6 +389,19 @@ export class UntitledTextEditorModel extends BaseTextEditorModel implements IUnt
...
+			return;
+		}
+
+		const lang = await this.languageDetectionService.detectLanguage(this.resource);
+		if (!lang) { return; }
+		this.setModeInternal(lang);
```
Comment (bpasero) [inline src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts:404]: Is it possible to run into race conditions here? E.g.:
- the model may be disposed already by the time this returns (e.g. editor closed)
- the model might have been changed and another language detection run occurs, should we wire through some kind of cancellation support?

File: src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts [hunk 12]
```diff
@@ -126,17 +131,21 @@ export class UntitledTextEditorModel extends BaseTextEditorModel implements IUnt
...
 
 		// Make known to working copy service
 		this._register(this.workingCopyService.registerWorkingCopy(this));
 
 		if (preferredMode) {
-			this.setMode(preferredMode);
+			this.setModeInternal(preferredMode);
 		}
 
+		this.initialMode = this.preferredMode ?? 'plaintext';
```
Comment (bpasero) [inline src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts:147]: There is a constant to use: `PLAINTEXT_MODE_ID`

File: src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts [hunk 13]
```diff
@@ -178,11 +187,23 @@ export class UntitledTextEditorModel extends BaseTextEditorModel implements IUnt
...
 
-	override setMode(mode: string, setExplicitly = true): void {
+	override setMode(mode: string): void {
+		if (!this._hasModeSetExplicitly
+			&& this.preferredMode
+			&& this.preferredMode !== this.initialMode
+			&& this.languageDetectionService.isEnabledForMode(this.preferredMode)
+		) {
+			// This is a best attempt to narrow it down to: "automatic language detection was wrong"
+			this.telemetryService.publicLog2('automaticlanguagedetection.likelywrong');
```
Comment (bpasero) [inline src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts:197]: I am not entirely sure this does something useful? If we are trying to detect the case of the auto detection mechanism doing something wrong, how is that related to the initial mode? Shouldn't we rather maybe track the user explicitly changing the mode from the status bar as a indication?

**Review by bpasero — CHANGES_REQUESTED** [review]

File: src/vs/workbench/services/languageDetection/browser/languageDetectionWorkerServiceImpl.ts (hunk 7 above)
Comment (TylerLeonhardt) [reply src/vs/workbench/services/languageDetection/browser/languageDetectionWorkerServiceImpl.ts]: take a look at the refactor in 3a8eb1a. I reuse a bunch more... it still feels a little weird though...

File: src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts (hunk 11 above)
Comment (TylerLeonhardt) [reply src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts]: > the model may be disposed already by the time this returns (e.g. editor closed)

I added a check around setting the mode that checks if it's been disposed.

> the model might have been changed and another language detection run occurs, should we wire through some kind of cancellation support?

Possible. Unfortunately, Tensorflow doesn't support cancellation natively... the best we can do is `Promise.race` with a token being cancelled. Is there prior art somewhere on that? I'd like to save this for later but I'll open an issue about it if that's ok with you.

File: src/vs/workbench/services/languageDetection/browser/languageDetectionWorkerServiceImpl.ts [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+		}
+		return this.worker;
+	}
+
+	override async _getProxy(): Promise<LanguageDetectionSimpleWorker> {
+		return await this._getOrCreateLanguageDetectionWorker().getProxyObject();
+	}
+
+	// foreign host request
+	public override fhr(method: string, args: any[]): Promise<any> {
```
Comment (isidorn) [inline src/vs/workbench/services/languageDetection/browser/languageDetectionWorkerServiceImpl.ts:157]: Why don't you make this method `async`, in that way you will not have to `Promise.resolve` wrap everything

File: src/vs/workbench/services/languageDetection/browser/languageDetectionWorkerServiceImpl.ts [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+	private worker: IWorkerClient<LanguageDetectionSimpleWorker> | undefined;
+
+	constructor(
+		modelService: IModelService,
+		private readonly _telemetryService: ITelemetryService,
+		private readonly _indexJsUri: string,
+		private readonly _modelJsonUri: string,
+		private readonly _weightsUri: string) {
+		super(modelService, true, 'languageDetectionWorkerService');
+
```
Comment (isidorn) [inline src/vs/workbench/services/languageDetection/browser/languageDetectionWorkerServiceImpl.ts:136]: Extra whitespace

**Comment by isidorn** [comment]
Cool PR! I did an initial review and left comments in the code.
Once you have something you are happy with let me know so I try this out end to end. Thanks 👏

File: src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts (hunk 13 above)
Comment (isidorn) [reply src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts]: Makes sense what Ben is saying. Also feel free to leave this telemetry work as a follow up.

File: extensions/vscode-api-tests/src/singlefolder-tests/untitled.languagedetection.test.ts [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+		"./typings",
+		"./vs"
+	]
+}`);
+		});
+
+		assert.ok(result);
+
+		// language detection is debounced so we need to wait a bit
+		await sleep(2000);
```
[the remark this answers is not in the document]
Comment (bpasero) [reply extensions/vscode-api-tests/src/singlefolder-tests/untitled.languagedetection.test.ts:71]: This is an integration test, but even there waiting 2s is ugly as it will make the builds slower.

File: src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts [hunk 17]
```diff
@@ -368,6 +375,24 @@ export class UntitledTextEditorModel extends BaseTextEditorModel implements IUnt
 
 		// Emit as general content change event
 		this._onDidChangeContent.fire();
+
+		this.autoDetectLanguage();
```
Comment (bpasero) [inline src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts:379]:
```suggestion
		// Try to detect language from content
		// (debounced by some time to reduce
		// pressure).
		this.autoDetectLanguage();
```

File: extensions/vscode-api-tests/src/singlefolder-tests/untitled.languagedetection.test.ts [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+
+import * as assert from 'assert';
+import * as vscode from 'vscode';
+import { assertNoRpc, closeAllEditors } from '../utils';
+
+function sleep(ms: number) {
+	return new Promise(resolve => setTimeout(resolve, ms));
+}
+
+suite('vscode - untitled automatic language detection', () => {
```
Comment (bpasero) [inline extensions/vscode-api-tests/src/singlefolder-tests/untitled.languagedetection.test.ts:14]: Isn't this test depending on language detection being enabled? If so should it be enabled as a workspace setting here: `extensions/vscode-api-tests/testWorkspace/.vscode/settings.json`

**Review by bpasero — CHANGES_REQUESTED** [review]
Found only minor things now on my end.

File: src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts (hunk 11 above)
Comment (bpasero) [reply src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts]: Yeah one ugly bit of the `@debounce` annotation is that we cannot clear the timeout when the model is disposed. That's why I typically prefer to manage my own `RunOnceScheduler` which is disposable.

File: src/vs/workbench/services/languageDetection/browser/languageDetectionWorkerServiceImpl.ts [hunk 19]
```diff
@@ -0,0 +1,209 @@
...
+	async detectLanguage(resource: URI): Promise<string | undefined> {
+		const language = await this._languageDetectionWorkerClient.detectLanguage(resource);
+		if (language) {
+			return this.getModeId(language);
+		}
+		return undefined;
+	}
+
+	async detectLanguages(resource: URI): Promise<string[]> {
```
[the remark this answers is not in the document]
Comment (TylerLeonhardt) [reply src/vs/workbench/services/languageDetection/browser/languageDetectionWorkerServiceImpl.ts:73]: Yes in the quick pick. I think it's still useful. If a text editor does contain 2 languages (I had an example where an editor contained both Java and C# code), showing both in the quick pick is convenient.

File: extensions/vscode-api-tests/src/singlefolder-tests/untitled.languagedetection.test.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (TylerLeonhardt) [reply extensions/vscode-api-tests/src/singlefolder-tests/untitled.languagedetection.test.ts:71]: Yeah I don't know how to do this correctly then because I need to wait for the debounce to finish (600ms).

Any ideas? Also this test is failing even though it worked locally so I'm guessing this sleep is the cause.

File: src/vs/workbench/services/languageDetection/browser/languageDetectionWorkerServiceImpl.ts (hunk 14 above)
Comment (TylerLeonhardt) [reply src/vs/workbench/services/languageDetection/browser/languageDetectionWorkerServiceImpl.ts:157]: Funny enough, copilot wrote this entire message and I glanced and said "yeah that looks right"...async/await is better here I think.

File: extensions/vscode-api-tests/src/singlefolder-tests/untitled.languagedetection.test.ts (hunk 18 above)
Comment (TylerLeonhardt) [reply extensions/vscode-api-tests/src/singlefolder-tests/untitled.languagedetection.test.ts:14]: Yeah in the test you can see its applying the LanguageDetection setting. However, your idea is better since having this setting on in other scenarios want do any harm.

File: extensions/vscode-api-tests/src/singlefolder-tests/untitled.languagedetection.test.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (TylerLeonhardt) [reply extensions/vscode-api-tests/src/singlefolder-tests/untitled.languagedetection.test.ts:71]: I was originally thinking there would be some kind of event that I could set up but that doesn't seem to be the case...unless I'm wrong.

File: src/vs/editor/common/services/editorSimpleWorker.ts [hunk 20]
```diff
@@ -326,13 +326,13 @@ declare const require: any;
 export class EditorSimpleWorker implements IRequestHandler, IDisposable {
 	_requestHandlerBrand: any;
 
-	private readonly _host: EditorWorkerHost;
+	protected readonly host: EditorWorkerHost;
```
[the remark this answers is not in the document]
Comment (TylerLeonhardt) [reply src/vs/editor/common/services/editorSimpleWorker.ts]: reverted.

File: src/vs/workbench/services/languageDetection/browser/languageDetectionWorkerServiceImpl.ts (hunk 14 above)
Comment (TylerLeonhardt) [reply src/vs/workbench/services/languageDetection/browser/languageDetectionWorkerServiceImpl.ts]: changed.

File: src/vs/workbench/browser/parts/editor/editorStatus.ts [hunk 21]
```diff
@@ -1284,6 +1291,28 @@ export class ChangeModeAction extends Action {
...
+				const guessRankOfPicked: number = (pick as IDetectedLanguageQuickPickItem).guessRank ?? -1;
+				// If we detected languages and they didn't choose the top detected language (which should also be the active language if automatic detection is enabled)
+				// then the automatic language detection was likely wrong and the user is correcting it. In this case, we want telemetry.
+				if (detectedLanguages.length && guessRankOfPicked !== 0) {
+					type AutomaticLanguageDetectionLikelyWrongData = {
+						choseOtherGuessedLanguage: boolean,
+						currentLanguageId: string,
+						nextLanguageId: string
+					};
+					type AutomaticLanguageDetectionLikelyWrongClassification = {
```
[the remark this answers is not in the document]
Comment (TylerLeonhardt) [reply src/vs/workbench/browser/parts/editor/editorStatus.ts]: moved into where the other interfaces are. Looks nice!

File: src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts (hunk 11 above)
Comment (TylerLeonhardt) [reply src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts]: I had no idea about `RunOnceScheduler`! that's nice!

File: src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts [hunk 22]
```diff
@@ -141,6 +143,8 @@ export class UntitledTextEditorModel extends BaseTextEditorModel implements IUnt
 			this.setModeInternal(preferredMode);
 		}
 
+		this._autoDetectLanguageScheduler = this._register(new RunOnceScheduler(() => this.autoDetectLanguage(), 600));
```
Comment (bpasero) [inline src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts:146]: Even more concise, you can move this entire line to the top where you declare the member.

File: extensions/vscode-api-tests/src/singlefolder-tests/untitled.languagedetection.test.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (TylerLeonhardt) [reply extensions/vscode-api-tests/src/singlefolder-tests/untitled.languagedetection.test.ts]: Figured it out.

**Comment by TylerLeonhardt** [comment]
merging this in now since I addressed all the feedback and I'd like to not maintain distro
